packages/flame/lib/src/game/transform2d.dart

```diff
@@ -0,0 +1,203 @@
+import 'dart:math' as math;
+import 'package:flutter/foundation.dart';
+import 'package:vector_math/vector_math_64.dart';
+import 'notifying_vector2.dart';
+
+/// This class describes a generic 2D transform, which is a combination of
+/// translations, rotations, reflections and scaling. These transforms are
+/// combined into a single matrix, that can be either applied to a canvas,
+/// composed with another transform, or used directly to convert coordinates.
+///
+/// The transform can be visualized as 2 reference frames: a "global" and
+/// a "local". At first, these two reference frames coincide. Then, the
+/// following sequence of transforms is applied:
+///   - translation to point [position];
+///   - rotation by [angle] radians clockwise;
+///   - scaling in X and Y directions by [scale] factors;
+///   - final translation by [offset], in local coordinates.
+///
+/// The class is optimized for repeated use: the transform matrix is cached
+/// and then recalculated only when the underlying properties change. Moreover,
+/// recalculation of the transform is postponed until the matrix is actually
+/// requested by the user. Thus, modifying multiple properties at once does
+/// not incur the penalty of unnecessary recalculations.
+///
+/// This class implements the [ChangeNotifier] API, allowing you to subscribe
+/// for notifications whenever the transform matrix changes. In addition, you
+/// can subscribe to get notified when individual components of the transform
+/// change: [position], [scale], and [offset] (but not [angle]).
+///
+class Transform2D extends ChangeNotifier {
+  final Matrix4 _transformMatrix;
+  bool _recalculate;
+  double _angle;
+  final NotifyingVector2 _position;
+  final NotifyingVector2 _scale;
+  final NotifyingVector2 _offset;
+  static const tau = 2 * math.pi;
+
+  Transform2D()
+      : _transformMatrix = Matrix4.identity(),
+        _recalculate = true,
+        _angle = 0,
+        _position = NotifyingVector2(),
+        _scale = NotifyingVector2()..setValues(1, 1),
+        _offset = NotifyingVector2() {
+    _position.addListener(_markAsModified);
+    _scale.addListener(_markAsModified);
+    _offset.addListener(_markAsModified);
+  }
+
+  factory Transform2D.copy(Transform2D other) => Transform2D()
+    ..angle = other.angle
+    ..position = other.position
+    ..scale = other.scale
+    ..offset = other.offset;
```

Comment:
I think you would want clones of these values and not the same references?

Reply:
A line `..position = other.position` invokes the setter
```
set position(Vector2 position) => _position.setFrom(position);
```
where `.setFrom` ultimately copies the values:
```
  void setFrom(Vector2 other) {
    final otherStorage = other._v2storage;
    _v2storage[1] = otherStorage[1];
    _v2storage[0] = otherStorage[0];
  }
```
So, in this case the code that I wrote should work just fine.

Also, I have a test `test('.copy', ...)` that verifies that a clone does not get modified when the original transform is modified.

Reply:
Aah, of course! Apparently I hadn't woken up yet when I did the review.